firedrake/preconditioners/pmg.py

```diff
@@ -151,7 +153,13 @@ def coarsen(self, fdm, comm):
         cele = self.coarsen_element(fV.ufl_element())
         cV = firedrake.FunctionSpace(fV.mesh(), cele)
         cdm = cV.dm
-        cu = firedrake.Function(cV)
+
+        if isinstance(fV.ufl_element(), MixedElement) and not isinstance(fV.ufl_element(), (VectorElement, TensorElement)):
+            cu = firedrake.Function(cV)
+            for (i, _) in enumerate(fV):
+                firedrake.interpolate(fu.split()[i], cu.split()[i])
+        else:
+            cu = firedrake.interpolate(fu, cV)
```

Comment:
Unfortunately this is not sufficient for nonlinear problems where the state continues to change (because you only ever inject this state once). So I think if you do:

```python
cu = firedrake.Function(cV)
interpolators = tuple(Interpolator(fus, cus) for fus, cus in zip(fu.split(), cu.split())
def inject_state(interpolators):
    for interpolator in interpolators:
        interpolator.interpolate()
```

And then after you've set up the context (line 208) do:

```python
add_hook(parent, setup=partial(inject_state, interpolators), call_setup=True)
```

I think you don't need a teardown because you're not actually stashing anything on a coarse dm.

Reply:
Aside, this issue is a problem for the general GMG code too (so if you do further h-multigrid, the coarsened things will only ever linearise around the initial state I fear).